### DISCUSSION TITLE
Partitioning and Ordering Information

### DISCUSSION THREAD

**Discussion by hawkfish** [duckdb/duckdb#13753] [Ideas]
## Overview

Partitioning and sorting are expensive operations.
Often this information can be reused by multiple operators in a pipeline.
The goal of this work is to find a way support such reuse
to both improve performance and reduce resource consumption during query processing.

## Partitioning Operators

Partitioned streams are naturally generated by some operators.
The window operator is the most obvious example, but there are others.
AsOf Joins partition on the equality keys to reduce the amount of sorting required.
Hash Grouping can produce disjoint partitions if threads take
ownership of particular keys during merge operations.

Operators that generate partitions can leverage incoming partitionings
if they are compatible with their own target partitionings.
In some situations, it may even be that the partitioning is _identical_,
but the operator needs to use it in a different manner. 

All operators that partition or order data are naturally sources.
This suggests that such operators should produce the same metadata as table scans.
It also suggests that table scans try to produce similar metadata when possible.

### Window Variants

There is a class of window functions called reducers
that have to be implemented in a single threaded manner because each value
depends on the one before. A new window operator to support these functions
might precede a traditional one, greatly reducing the partitioning and sorting
load on the second. 

Or some functions maybe partitioned at a coarser level of detail,
which would require two operators, but one could reuse the existing partitioning 
to reduce the work.  

A third windowing example is the streaming window operator.
If some functions are compatible with streaming except for partitioning,
then a partition-aware version of the operator could process partitions
sequentially, resetting the function states on partition boundaries.
This could even be made multi-threaded, with different partitionings 
running on separate threads.

### Aggregation Variants

The first relational systems used sorting for many operations,
including aggregation. The advantage was that it made aggregation 
into a streaming operator, but the sorting cost was prohibitive,
and this approach was widely replaced by hash aggregation.

Streaming aggregation does not actually require a full sort; 
partitioning on the aggregation keys is sufficient.
If the tuple stream is partitioned on the grouping keys, 
then partitioned can be implemented by simply looking for key 
boundaries. 

A further generalisation comes by observing that
if the stream is partitioned on a subset of the keys,
then a hash table can be built on the unpartitioned keys alone
and flushed on partition boundaries. This reduces the size of
the hash table both horizontally and vertically (reducing memory footprint)
while allowing the operator to still stream the tuples with some buffering.

### Join Variants

Historically, relational systems also used sorting for joins, 
until the advent of hash join algorithms.
Similarly, the advantage of merge joins is that they can stream
both sides of the join if both sides are sorted on the join keys.
Outer join semantics can also be implemented as part of the streaming
by emitting the appropriate tuple whenever a sort key changes on either side.
This also greatly reduces the memory footprint by removing a potentially large
piece of bookkeeping.

For multiple inequalities, our existing IEJoin algorithm sorts both sides
of the join and applies a dual sort matching algorithm. This is an expensive
operation, and avoiding it would allow us to use the algorithm at a much lower cost. 

### Threading

A common theme with partitioning is that it is most effective if the
partitions are on separate threads. For performance and resource management
reasons, some operators (such as Window) may emit partitions on multiple threads.
We may find it helpful to create a lightweight Exchange-type operator that can
reroute partitions to individual threads, or to enhance the pipeline architecture
to handle such issues (similar to block compression)

## Metadata Components

At present, we support the notion of a `batch index`,
which is essentially the the morsel number from the scan.
This enables streaming operators that can preserve order to
tag their vector stream for reassembly without an explicit sort.
We would like to extend this to include other partitioning and ordering metadata.
This would allow operators to cleanly leverage existing 
partitioning and ordering.

One complicating factor is that there may be _multiple_ descriptions
of a stream's partitioning and ordering due to functional dependencies.
Typically these require extra work to derive, but it may be necessary 
to model them if they fall out naturally (e.g., from an aggregation operator).

Partitioned and ordered streams will naturally be split on partition boundaries
and then major sort boundaries.

### Statistics

We currently have a notion of column statistics that is global to the stream.
But there is no reason why this data structure could not be used at a finer
level of detail. Categorical domains, value ranges (zone maps), and NULLification
are all valuable metadata that can be generated by any source operator,
not just table scans. It seems reasonable to build on this existing machinery
to implement partitioning and sorting metadata.

### Partitioning

Partitioning can generated from sorting or hashing, but for the purposes of this discussion
we shall restrict ourselves to the hashing situation. Note that
"hashing" is a somewhat loose term and can also be applied to physical partitions 
on disk such as hive partitions (one can take the view that inodes are hashes of
the partition keys...)

Partition keys are effectively unordered, so they can be stored in any order
as long as comparisons are also unordered. They can also be converted to a lower
level of detail by removing key(s).

Since the values in a partition are not likely to be adjacent,
the natural representation of the domain of a partition key appears to be
categorial statistics rather than ranges. In practice, however, it may be
that the smallest cardinality key(s) would use this representation and
the the higher ones call back to ranges. 

### Ordering

Orderings are generally created by sorting operators, but may result from
data that is "naturally" ordered. The latter situation would require 
metadata to be stored with the underlying data. Sorting properties are so
valuable, however, that detecting them on read and writing them back for future
use (e.g., in a db files) may be worth doing on the first read, or even first write.

Ordering keys are nested, so they can only be removed from the right.
They can be converted to partitions on most significant key boundaries,
however, at which point the other keys can be removed as needed.

The values of an ordering are likely to be adjacent, especially the outer key.
This suggests that the default representation for the domain would be a range statistic.

## Metadata Proposal

As described above, the current runtime metadata model is a `block index`.
We propose that this be converted to a structure containing the index,
along with one or more partition and ordering descriptions.
A partition description would be a set of partitioning attributes for the entire stream
and specific statistics for the current block.
An ordering description would a list of sort keys (including sense and NULL semantics)
and the specific statistics for the current block.

Sources can optionally produce these descriptions at runtime, much as they produce
block indexes today. It may be that an operator has to create its own block indices, 
e.g., Window might choose to index blocks based on group hash and sort block, which would be
deterministic, but unrelated to the scan indicies. The only requirement is that
the reassembled stream using indicies should be partitioned and sorted as described
by the metadata keys.

## References

[1] Cao et al., [_Optimization of Analytic Window Functions_](https://vldb.org/pvldb/vol5/p1244_yucao_vldb2012.pdf), VLDB 2012
[2] Do et al., [_Efficient Sorting, Duplicate Removal, Grouping, and Aggregation_](https://dl.acm.org/doi/pdf/10.1145/3568027), ACM Transactions on Database Systems, Volume 47, Issue 4.

**Comment by lnkuiper** [comment]
This sounds great! Thanks for writing your thoughts down. I wonder what the most common cases are, where the biggest gains are, and what we should try tackling first. Do we know of any benchmarks that would benefit from this so that we have a target? For hive-partitioned Parquet data, it would already help on TPC-H, for example.

**Reply by hawkfish** [reply]
Yikes I left out one of the most important use cases: avoiding sorting!  As mark put it

>  I think you can also exploit this for speeding up sorting
> e.g. if you know that row group 1 has dates in the range of [2020-01-01 - 2020-01-31] and row group 2 has dates in the range of 
> [2020-02-01 - 2020-02-28] you can sort the two ranges independently and skip the merging step

**Comment by pthatte1-bb** [comment]
Is there an issue related to this that we can watch?

> I wonder what the most common cases are, where the biggest gains are, and what we should try tackling first.

 Just one example - for interactive analytics on constituents of SPY index: ordering-within-partitions is needed (e.g. forward-fill/back-fill before time-series-analytics), but ordering across partitions is not.

**Reply by hawkfish** [reply]
This is the issue, so watch this space.

Ordering across partitions is not usually what one wants to exploit but rather leveraging partitioning between operators (e.g., window operators as in reference [1]) AsOf joins also use partitioning for equality predicates (and IEJoin maybe should too) but there is no need for ordering there.  All of these operators are used for temporal operations, so they could all benefit.

**Reply by pthatte1-bb** [reply]
> and IEJoin maybe should too

+1

**Comment by bskubi** [comment]
Just wanted to chime in with encouragement on this feature, especially if taking advantage of sorting information could be used to speed up selecting small subsets of the table from the major sort column!

In the bioinformatics community, we lean heavily on presorted data to make our algorithms and queries efficient. Our datasets are getting very large and granular with single cell data. Currently, even writing to a partitioned parquet file can't quite match the query performance or compactness of my field's existing HDF5-based format for storing this data. Using the HDF5 based format means we miss out on the expressiveness of SQL, but query performance matters a lot too due to the large amount of machine learning work going on in this area.

I'll keep an eye on this space!
